### ISSUE TITLE
Cloudflare subdomain issue

### ISSUE THREAD

**Issue by mikopia** [rmbolger/Posh-ACME#184]
I am trying to generate a certificate for a sub domain of a domain i have in Cloudflare.

Domain: domain.com
Subdomain: app1.domain.com

The problem is that the cloudflare plugin can't find the zone called app1.domain.com, so it fails.

```
PS C:\Users\Administrator> New-PACertificate app1.domain.com
 -DnsPlugin Cloudflare -PluginArgs $pArgs
Unable to find Cloudflare hosted zone for _acme-challenge.app1.domain.com
At C:\Users\Administrator\Documents\WindowsPowerShell\Modules\Posh-ACME\3.8.0\DnsPlugins\Cloudflare.ps1:103 char:9
+         throw "Unable to find Cloudflare hosted zone for $RecordName"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Unable to find app1.domain.com
:String) [], RuntimeException
    + FullyQualifiedErrorId : Unable to find Cloudflare hosted zone for _acme-challenge.app1.domain.com

```
I overrode this by specifying a zoneID in cloudflare.ps1 and the rest worked.

**Comment by rmbolger** [comment]
Which type of credential are you using?

**Comment by MetUys** [comment]
Odd I had no issue with this an even deeper use... app1.sub1.domain.ccTLD
my code:
`New-PACertificate app1.sub1.domain.ccTLD -AcceptTOS -Contact $Email -DnsPlugin Cloudflare -PluginArgs @{CFAuthEmail=$Email; CFAuthKey='ohnoyoudont'} -PfxPass $CertsPassword -DnsSleep 5 -Install -Verbose`

**Comment by Jackbennett** [comment]
With the api token set to specifically my example.co.uk zone Find-CFzone errors;

```powershell
Invoke-RestMethod : The remote server returned an error: (403) Forbidden.
At C:\Program Files\WindowsPowerShell\Modules\posh-acme\3.8.0\DnsPlugins\CloudFlare.ps1:253 char:27
+ ...    $result = Invoke-RestMethod "$($apiRoot)?page=$page&per_page=50"
```

If I hit `https://api.cloudflare.com/client/v4/zones` manually when my API token is set to a specific zone I get a 403

looking at the code and other issues online you know about this, I wonder if CF just changed the permissions again for the api because when using the zone identifier I get make zone requests.

I don't see how to get the ZoneID From the cloudflare Dashboard UI. `https://dash.cloudflare.com/e.../DNS/etc` from the dashboard URL appears to be some account ID not the domain zone ID.

**Comment by rmbolger** [comment]
I think ultimately the problem is that the workaround I added for the limited token API bug now requires you to associated the token's permissions with all zones on the account rather than a limited set of zones in order to be able to effectively "list all zones". I still need to test to make sure, but that's my guess at the moment.

If that's the case, I may end up reverting that change and just deal with the 403 error using the old way. Maybe I'll throw a warning. The code won't be able to definitively know whether there's actually a permissions problem or that the zone it's checking for doesn't exist. But that's probably a better trade-off than forcing you to give the token more permissions than it actually needs. And then if Cloudflare ever actually fixes the bug, the warnings will hypothetically just go away.

As far as looking up the zone ID, I see both the zone ID and the account ID in the bottom right hand side of the domain's overview page. But since plugin arguments are saved per-account, I can't just add the zoneID as a plugin argument because it would change for different certs in different zones and renewals would break.

**Comment by rmbolger** [comment]
It looks like I was wrong and it's worse than I thought. Apparently if you don't give the token read access to all zones on the account, all calls to the "list zones" fail with an HTTP 403 even if you try to limit the results to only a zone you have been given access to. So there's basically no way for that token to discover a zone's ID.

So there are basically two options I can see.
- Force users to give the token access to all zones (which implies Edit to all zones as well since you can't differentiate different permissions scopes on a single token).
- Add the ability to use a second read-only token that only has read access to all zones. The primary token can still have edit limited to specific zones. But the plugin will use the read-only token to discover the Zone IDs. It's ugly, but it would work.

**Comment by Jackbennett** [comment]
>As far as looking up the zone ID, I see both the zone ID and the account ID in the bottom right hand side of the domain's overview page. But since plugin arguments are saved per-account, I can't just add the zoneID as a plugin argument because it would change for different certs in different zones and renewals would break.

You need to make a Cloudflare plugin object for each domain to sign, it's just convenient the module re-uses the plugin I only specify one. So you can keep the CFToken and CFZoneID together. I shouldn't think zoneID changes between accounts.

Ahh I see the zoneID now I wasn't scrolling down there.

So you could use the ID if specified, try listing the zone if not and show the correct error if there's a 403 about token scope or set the ID.

**Comment by rmbolger** [comment]
> You need to make a Cloudflare plugin object for each domain to sign, it's just convenient the module re-uses the plugin I only specify one.

Unfortunately, making that work would be a fairly large architectural change to the module. However, there will very likely be an option that could enable this scenario in 4.0. I'm going to have to implement something similar for HTTP plugins and I might as well extend that functionality to DNS plugins as well.

I think the optional secondary read-only token is the path I'm going to take on this for now. So your options for using a limited scope token will either be:
- A single token that has `Zone - Read` and `DNS - Edit` for all zones.
- One token that has `DNS - Edit` on a subset of zones and one token that has `Zone - Read` for all zones.

**Comment by mikopia** [comment]
What if the zoneid could be passed as a parameter?
________________________________

**Comment by rmbolger** [comment]
> What if the zoneid could be passed as a parameter?

That's basically what @Jackbennett was asking for. The plugin parameters are saved per-account rather than per-order. So it would only work if the all the cert orders associated with an ACME account have names in the same underlying domain. While this might be ok for some, it definitely wouldn't be for everyone.

The root of the problem here is that the Cloudflare API doesn't have a way to look up a zone ID based on a zone name with a token that doesn't have at least read access to all zones on an account. It's not the end of the world and it's still far better than needing to supply the Global API key which has full access to everything on the account. It's just a bit more effort to generate a second token if you want to give the module the absolute minimum possible permissions.

**Comment by mikopia** [comment]
Yes, but what if the workaround was to provide the zoneid as a parameter every time it was called? Or it cannot be saved with the order, the same way that you could have multiple certificates with different api keys on different accounts?
________________________________

**Comment by rmbolger** [comment]
I mean, I suppose it could be yet another optional parameter with all the caveats listed above. But the different parameter permutations are already getting somewhat complicated. Is it really important enough just to avoid making a second token?

**Comment by mikopia** [comment]
No, I suppose the 2 token way works too.
________________________________

**Comment by rmbolger** [comment]
Ok. The `CFToken` and `CFTokenInsecure` parameters now have matching optional `CFTokenReadAll` and `CFTokenReadAllInsecure` parameters that will be used to do the zone ID lookups if specified. I also updated the user guide to describe the additional options.

Try out the dev/master version and let me know if this works for you guys.

**Comment by mikopia** [comment]
nope, not working.

$CFTokenSecure = Read-Host -AsSecureString -Prompt 'CFToken'
$CFTokenReadAllSecure = Read-Host -AsSecureString -Prompt 'CFTokenReadAll'
$pArgs = @{ CFToken = $CFTokenSecure; CFTokenReadAll = $CFTokenReadAllSecure  }

New-PACertificate app1.domain.com -DnsPlugin Cloudflare -PluginArgs $pArgs

VERBOSE: Publishing DNS challenge for app1.domain.com
VERBOSE: Attempting to find hosted zone for _acme-challenge.app1.domain.com
VERBOSE: GET https://api.cloudflare.com/client/v4/zones/?name=app1.domain.com with 0-byte payload
WARNING: There was a permissions error checking the existence of app1.domain.com. This either indicates the zone
doesn't exist or the supplied credentials are invalid. If this is the domain apex and you know the zone exists, check
your token permissions. Otherwise, ignore this message.
VERBOSE: GET https://api.cloudflare.com/client/v4/zones/?name=domain.com with 0-byte payload
WARNING: There was a permissions error checking the existence of domain.com. This either indicates the zone doesn't
exist or the supplied credentials are invalid. If this is the domain apex and you know the zone exists, check your
token permissions. Otherwise, ignore this message.
Unable to find Cloudflare hosted zone for _acme-challenge.app1.domain.com
At C:\Users\Administrator\Documents\WindowsPowerShell\Modules\Posh-ACME\DnsPlugins\Cloudflare.ps1:30 char:9
+         throw "Unable to find Cloudflare hosted zone for $RecordName"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Unable to find ...ifi.domain.com:String) [], RuntimeException
    + FullyQualifiedErrorId : Unable to find Cloudflare hosted zone for _acme-challenge.app1.domain.com

**Comment by nebula-it** [comment]
Hey @rmbolger ,
I'm trying to use the two token system. So in 
```
$token = Read-Host -AsSecureString -Prompt 'API Token'
$pArgs = @{ CFToken = $token }
# (Optional) Only specify the ReadAll token if you generated one
$pArgs.CFTokenReadAll = Read-Host -AsSecureString -Prompt 'Secondary Token'
```
The "API Token" is the one that has DNS Edit permissions for the specific Zone and "Secondary Token" is the one that has some Zone Read permissions on all the zones?

I have been trying this but still get error `Unable to find Cloudflare hosted zone for _acme-challenge.mydomain.com`

If I test the Zone Read token in 
```
$header = @{Authorization="Bearer <token value>"}
irm https://api.cloudflare.com/client/v4/zones/?name=domain.com -Headers $header
```
It gets authenticated and I get the results.
Any ideas?

**Comment by rmbolger** [comment]
Hi @nebula-it. The ability to use the two-token system hasn't been added to an official release yet (despite the guide being updated as if it had). Are you using the dev/master branch to test with or the latest 3.8.0 release from the gallery?

If you *are* using the dev version, can you start a fresh PowerShell session, run the following, and send me the output (z97y9drumh (at) snkmail.com)?

```powershell
$token = Read-Host -AsSecureString -Prompt 'API Token'
$token2 = Read-Host -AsSecureString -Prompt 'Secondary Token'
$pArgs = @{ CFToken = $token; CFTokenReadAll = $token2 }
$DebugPreference = 'Continue'
Publish-DnsChallenge domain.com (Get-PAAccount) faketoken Cloudflare $pArgs -Verbose
```

**Comment by nebula-it** [comment]
Hey @rmbolger . Thank you for quick reply. Yes I figured it as much that this change was not yet pushed into official release. 
So I was trying the dev/master branch by downloading the zip. But I'm unable to import module, it fails with following error:
```
PS C:\WINDOWS\system32> Import-Module Posh-ACME
Import-Module : Could not load file or assembly 'file:///<path to module>\Posh-ACME\lib\BouncyCastle.Crypto.dll' or one of its dependencies.
Operation is not supported. (Exception from HRESULT: 0x80131515)
At line:1 char:1
+ Import-Module Posh-ACME
+ ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Import-Module], FileLoadException
    + FullyQualifiedErrorId : FormatXmlUpdateException,Microsoft.PowerShell.Commands.ImportModuleCommand
```
Going by other threads it seems like .NET 4.7.1 is required but I already have it installed. And I tested this on two different system with Windows version 1809 and 1903, powershell version 5.1.17763.771 and 5.1.18362.145 respectively.
Maybe I should open a new issue for this 😬

**Comment by rmbolger** [comment]
If you downloaded the zip file with a browser, it's probably tagged with the "from the Internet" flag which gets passed to anything extracted from it and PowerShell is trying to protect you from loading potentially unsafe DLLs. Open the properties dialog for the zip file and look down near the bottom for a check box to "Unblock" the file. Check the box click OK, then delete the files you extracted, re-extract a new set, and try importing the module again. Alternatively, I think you can also just Unblock the DLL file itself which is in the `lib` folder inside the module.

**Comment by nebula-it** [comment]
Ah yeah, that got it going. Thank you so much.
Any idea when this will be pushed into PS gallery?

**Comment by rmbolger** [comment]
Ultimately, I was just waiting on someone other than me to confirm the two-token code is working.

**Comment by nebula-it** [comment]
Well I can confirm it works perfectly fine with "Seconday Token" with all Zone Read permissions. So I think you can push it.
I'm deploying this to get certs for all our internal services sometime next week so it would be lovely if I can just pull this version from PS gallery.
And thank you for your work, this is really elegant work.

**Comment by rmbolger** [comment]
Works for me. 3.9.0 is now available.

**Comment by nebula-it** [comment]
Hey @rmbolger ,
Just another thought on this, one other option can be to use the zoneID directly in the script. I haven't gotten a chance yet to dig deeper into Cloudflare API but you might know it, is something like this possible?
The reason being our team want the minimum possible permissions used so if we can replace the 'Read All Zone' API Token with a 'Zone ID' for that particular zone. I'll try to see if I can figure it out as well but posting here to hear your thoughts on this.

**Comment by rmbolger** [comment]
We actually discussed this a bit earlier in this issue thread. Short version is that the way the plugin architecture works right now, there's no good way to add per-order plugin arguments like a zone ID. It would have to be per-account. So you're either constantly overwriting the zone ID for different orders and you have to deal with renewals manually or you need to make a separate account per order.

What I'd really love is for Cloudflare to just fix their API so that there's a standard way to query a zone ID based on the domain name that only requires giving access to the zone you care about. All of this hassle is just working around that limitation.

**Comment by nebula-it** [comment]
Thanks for the explanation. I'll try to contact cloudflare support on this and see if they have plans to fix this.
Cheers
